flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManager.java

```diff
@@ -18,81 +18,267 @@
+		this.closed = false;
+
+		for (File localStateRecoveryRootDir : localStateRootDirectories) {
+			if (!localStateRecoveryRootDir.exists()) {
+
+				if (!localStateRecoveryRootDir.mkdirs()) {
+					throw new IllegalStateException("Could not create root directory for local recovery: " +
+						localStateRecoveryRootDir);
+				}
+			}
+		}
+
+		// install a shutdown hook
+		this.shutdownHook = new Thread("TaskExecutorLocalStateStoresManager shutdown hook") {
+			@Override
+			public void run() {
+				shutdown();
+			}
+		};
+		try {
+			Runtime.getRuntime().addShutdownHook(this.shutdownHook);
+		} catch (IllegalStateException e) {
+			// race, JVM is in shutdown already, we can safely ignore this
+			LOG.debug("Unable to add shutdown hook for TaskExecutorLocalStateStoresManager, shutdown already in progress", e);
+		} catch (Throwable t) {
+			LOG.warn("Error while adding shutdown hook for TaskExecutorLocalStateStoresManager", t);
+		}
 	}
 
-	public TaskLocalStateStore localStateStoreForTask(
-		JobID jobId,
-		JobVertexID jobVertexID,
-		int subtaskIndex) {
+	@Nonnull
+	public TaskLocalStateStore localStateStoreForSubtask(
+		@Nonnull JobID jobId,
+		@Nonnull AllocationID allocationID,
+		@Nonnull JobVertexID jobVertexID,
+		@Nonnegative int subtaskIndex) {
+
+		synchronized (lock) {
+
+			if (closed) {
+				throw new IllegalStateException("TaskExecutorLocalStateStoresManager is already closed and cannot " +
+					"register a new TaskLocalStateStore.");
+			}
+
+			final Map<JobVertexSubtaskKey, TaskLocalStateStoreImpl> taskStateManagers =
+				this.taskStateStoresByAllocationID.computeIfAbsent(allocationID, k -> new HashMap<>());
+
+			final JobVertexSubtaskKey taskKey = new JobVertexSubtaskKey(jobVertexID, subtaskIndex);
 
-		Preconditions.checkNotNull(jobId);
-		final JobVertexSubtaskKey taskKey = new JobVertexSubtaskKey(jobVertexID, subtaskIndex);
+			// create the allocation base dirs, one inside each root dir.
+			File[] allocationBaseDirectories = allocationBaseDirectories(allocationID);
 
-		final Map<JobVertexSubtaskKey, TaskLocalStateStore> taskStateManagers =
-			this.taskStateManagers.computeIfAbsent(jobId, k -> new HashMap<>());
+			LocalRecoveryDirectoryProviderImpl directoryProvider = new LocalRecoveryDirectoryProviderImpl(
+				allocationBaseDirectories,
+				jobId,
+				allocationID,
+				jobVertexID,
+				subtaskIndex);
 
-		return taskStateManagers.computeIfAbsent(
-			taskKey, k -> new TaskLocalStateStore(jobId, jobVertexID, subtaskIndex));
+			LocalRecoveryConfig localRecoveryConfig = new LocalRecoveryConfig(
+				localRecoveryMode,
+				directoryProvider);
+
+			return taskStateManagers.computeIfAbsent(
+				taskKey,
+				k -> new TaskLocalStateStoreImpl(
+					jobId,
+					allocationID,
+					jobVertexID,
+					subtaskIndex,
+					localRecoveryConfig,
+					discardExecutor));
+		}
 	}
 
-	public void releaseJob(JobID jobID) {
+	public void releaseLocalStateForAllocationId(@Nonnull AllocationID allocationID) {
 
```

Comment:
In `TaskExecutorLocalStateStoresManager` I only found this `releaseLocalStateForAllocationId ()` which releases resources according to `allocationID`. Is there missing a function for releasing resources according to `jobID`? I think it should be used when a job is cancelled.